doc/manual/using_drtio_subkernels.rst

```diff
@@ -0,0 +1,206 @@
+
+Distributed Direct Memory Access (DDMA)
+---------------------------------------
+
+By default on DRTIO systems, all events recorded by the master's DMA core are kept and played back on the master. With distributed DMA, RTIO events that should be played back on remote destinations are distributed to the corresponding satellites. In some cases (typically, large buffers on several satellites with high event throughput), it allows for better performance and higher bandwidth, as the RTIO events do not have to be sent over the DRTIO link(s) during playback.
+
+To enable distributed DMA for the master, simply provide an ``enable_ddma=True`` argument for the :meth:`~artiq.coredevice.dma.CoreDMA.record` method - taking a snippet from the non-distributed example in the :ref:`core language tutorial <getting-started-dma>`: ::
+
+        @kernel
+        def record(self):
+            with self.core_dma.record("pulses", enable_ddma=True):
+                # all RTIO operations now go to the "pulses"
+                # DMA buffer, instead of being executed immediately.
+                for i in range(50):
+                    self.ttl0.pulse(100*ns)
+                    delay(100*ns)
+
+In standalone systems, as well as in subkernels (see below), this argument is ignored; in standalone systems it is meaningless and in subkernels it must always be enabled for structural reasons.  
+
+Enabling DDMA on a purely local sequence on a DRTIO system introduces an overhead during trace recording which comes from additional processing done on the record, so careful use is advised. Due to the extra time that communicating with relevant satellites takes, an additional delay before playback may be necessary to prevent a :exc:`~artiq.coredevice.exceptions.RTIOUnderflow` when playing back a DDMA-enabled sequence.
+
+Subkernels
+----------
+
+Rather than only offloading the RTIO channels to satellites and limiting all processing to the master core device, it is fully possible to run kernels directly on satellite devices. These are referred to as *subkernels*. Using subkernels to process and control remote RTIO channels can free up resources on the core device.  
+
+Subkernels behave for the most part like regular kernels; they accept arguments, can return values, and are marked by the decorator ``@subkernel(destination=i)``, where ``i`` is the satellite's destination number as used in the routing table. To call a subkernel, call it like any other function. There are however a few caveats: 
+
+   - subkernels do not support RPCs,
+   - subkernels do not support (recursive) DRTIO (but they can call other subkernels and send messages to each other, see below),
```

Comment:
I gather (though I didn't check) that DRTIO is technically possible specifically within DMA operations -- since it's fed back through the master, and it's explicitly mentioned in one of the implementation PRs that this will be distributed among satellites as necessary -- but I deliberately didn't mention this as it seems like something of more of a coincidence than a feature, and probably unnecessarily confusing in this context.

Reply:
"Does not support DRTIO" is taken directly from the original phrasing. #2235 describes it the same way. The addition of "recursive" is meant to clarify that they are in themselves DRTIO, but don't support _further_ DRTIO, so to speak. I understood it to mean subkernels are not allowed to access RTIO channels of other satellites / the master... except for the apparent loophole via DDMA, unless that's blocked off by some intermediate mechanism ....? @Spaqin?

Reply:
DDMA is done differently - rather than just directly playing back all RTIO events, they get sorted and sent out to appropriate satellites to be played back later. Master is one exception to this, there's no support in firmware for playing back DDMA. 

Which means DDMA is using the aux protocol with routing that allows the aux packets to go anywhere, rather than the main DRTIO protocol. So yes, there would be a "loophole" in that sense.

I am quite sure that it wouldn't be possible to dispatch events upstream, but downstream (to further satellites) may still work, that should be verified.

Reply:
Right, so the question to me remains whether to describe this DDMA quote-unquote "loophole" in the docs or not. Or I can also just rephrase if we don't like calling this "recursive DRTIO", which admittedly is not very clear

Reply:
> Why not?

DMA control is accessible in the [kernel CPU only](https://github.com/m-labs/artiq/blob/master/artiq/gateware/targets/kasli.py#L319) on standalone/master firmware. Asking to do it from comm core (where all DDMA data is sent and handled) in current architecture (where kernel CPU sends requests and comm only replies) wouldn't be possible, and frankly it's not a huge disadvantage - if you need DDMA involving master, just run it from the master.

Reply:
> But then it's not distributed since all events have to come from the master's SDRAM and RTIO CRI bus.

No, it can be distributed still (distribution must be [enabled first](https://github.com/m-labs/artiq/blob/master/artiq/coredevice/dma.py#L78)).  Just the local part runs from kernel CPU.

 > I suppose it's the same on satellites, subkernels cannot start DMA?
 
 They can, but it's a workaround - DMA start syscall is different for satellites, and it [sends a request](https://github.com/m-labs/artiq/blob/master/artiq/firmware/ksupport/lib.rs#L451) to the comm CPU to start DMA. On master, similar approach could be used, but since kernel<->comm communication is slower, I don't think that's a good approach, since it would slow down every DMA call, regardless if DDMA is used or not.